### PR TITLE
[add-const] do not recompute the uctx for the evar map

### PR DESCRIPTION
**Pull request by gares**
https://github.com/coq/coq/pull/19336

### CODE REVIEW

**Comment by gares** [comment]
CC @ppedrot

**Comment by gares** [comment]
Hello, I need help in order to remove the use of the non-sound API.
I'm getting this:
```
Debug: elpi lets escape exception: Conversion test raised an anomaly:
Anomaly
"in Univ.repr: Universe elpi.apps.locker.tests.test_locker.93 undefined."
Please report at http://coq.inria.fr/bugs/.
File "./apps/locker/tests/test_locker.v", line 95, characters 0-79:
Error: Conversion test raised an anomaly:
Anomaly
"in Univ.repr: Universe elpi.apps.locker.tests.test_locker.93 undefined."
Please report at http://coq.inria.fr/bugs/.
Raised at Exninfo.iraise in file "clib/exninfo.ml", line 79, characters 4-11
Called from AcyclicGraph.Make.enforce_lt in file "lib/acyclicGraph.ml", line 670, characters 15-28
Called from UGraph.enforce_constraint0 in file "kernel/uGraph.ml", line 94, characters 38-68
Called from UGraph.enforce_constraint in file "kernel/uGraph.ml", line 100, characters 37-62
Called from Stdlib__Set.Make.fold in file "set.ml", line 383, characters 34-55
Called from UState.merge_constraints in file "engine/uState.ml", line 286, characters 6-38
Called from UState.add_universe_constraints in file "engine/uState.ml", line 714, characters 16-60
Called from Evd.add_universe_constraints in file "engine/evd.ml", line 962, characters 23-68
Called from Reductionops.infer_conv_gen in file "pretyping/reductionops.ml", line 1320, characters 17-58
Re-raised at Exninfo.iraise in file "clib/exninfo.ml", line 81, characters 4-38
Called from Evarconv.infer_conv_noticing_evars in file "pretyping/evarconv.ml", line 629, characters 8-70
Called from Evarconv.evar_conv_x in file "pretyping/evarconv.ml", line 646, characters 6-80
Called from Evarconv.unify_leq_delay in file "pretyping/evarconv.ml", line 1959, characters 8-45
Called from Typing.judge_of_apply.apply_rec in file "pretyping/typing.ml", line 105, characters 10-58
Called from Typing.type_of in file "pretyping/typing.ml", line 646, characters 17-36
```
that comes from
```coq
mlock #[universes(polymorphic)] Definition id2@{u} (T : Type@{u}) (x : T) := x.
```
Note that f4340e1ae8c422d9ba51abbfb6536c2908f0a94a works if you run `dune build apps/locker/tests/test_locker.vo`

**Comment by gares** [comment]
The message above is for @SkySkimmer and @ppedrot

File: src/coq_elpi_builtins.ml [hunk 1]
```diff
@@ -941,6 +936,23 @@ let mkCLocalAssum x y z = Constrexpr.CLocalAssum(x,None,y,z)
...
+[%%elif coq = "8.20"]
+let declare_definition hack using ~cinfo ~info ~opaque ~body sigma =
+  let using = Option.map Proof_using.using_from_string using in
+  let gr = Declare.declare_definition ~cinfo ~info ~opaque ~body ?using sigma in
+  gr, Option.get !hack
+[%%else]
+let declare_definition _ using ~cinfo ~info ~opaque ~body sigma =
+  let using = Option.map Proof_using.using_from_string using in
+  let gr, uctx = Declare.declare_definition_full ~cinfo ~info ~opaque ~body ?using sigma in
+  gr, UState.context_set uctx
```
Comment (SkySkimmer) [inline src/coq_elpi_builtins.ml:952]: maybe need to do `if poly then empty else UState.context_set uctx` (not sure where `poly` is, I think either `cinfo` or `info`)

File: src/coq_elpi_HOAS.ml [hunk 2]
```diff
@@ -882,16 +882,24 @@ let show_coq_engine ?with_univs e = Format.asprintf "%a" (pp_coq_engine ?with_un
...
 
- let from_env_keep_univ_and_sigma ~env0 ~env sigma0 =
+
+[%%if coq = "8.19" || coq = "8.20"]
+let demote uctx sigma0 env = 
+      let uctx = UState.update_sigma_univs (Evd.evar_universe_context sigma0) (Environ.universes env) in
+      UState.demote_global_univs env uctx
+[%%else]
+ let demote uctx sigma0 env =
+   UState.demote_global_univs_sound uctx (Evd.evar_universe_context sigma0)
```
Comment (SkySkimmer) [inline src/coq_elpi_HOAS.ml:892]: Might be useful to add a debug print like "demoting (print uctx)"

File: .nix/config.nix [hunk 3]
```diff
@@ -32,7 +32,13 @@
       reglang.job = false;
     };
 
+    "coq-master".ocamlPackages = {
```
Comment (gares) [inline .nix/config.nix:35]: remove before merge

File: src/coq_elpi_builtins.ml [hunk 4]
```diff
@@ -534,11 +539,11 @@ let is_global_level env u =
 [%%endif]
 
 let err_if_contains_alg_univ ~depth t =
-  let global_univs = UGraph.domain (Environ.universes (Global.env ())) in
+  let env = Global.env () in
   let is_global u = 
     match Univ.Universe.level u with
     | None -> true
-    | Some l -> Univ.Level.Set.mem l global_univs in
+    | Some l -> is_global_level env l in
```
Comment (SkySkimmer) [inline src/coq_elpi_builtins.ml:546]: you could merge this in master separately btw

Comment (gares) [reply src/coq_elpi_builtins.ml:546]: right, but I hope to merge this soon, keeping the PR going is draining too much of my time

**Comment by ppedrot** [comment]
Please merge now

**Pull request merged**